sdk/core/System.ClientModel/src/Pipeline/MessageLoggingPolicy.cs

```diff
@@ -0,0 +1,342 @@
+        }
+
+        private void DecrementLength(ref int count)
+        {
+            var left = Math.Min(count, _bytesLeft);
+            count = left;
+
+            _bytesLeft -= count;
+        }
+    }
+    #endregion
+
+    #region LoggingStream
+    private class LoggingStream : Stream
+    {
+        private readonly string _requestId;
+        private int _maxLoggedBytes;
+        private readonly int _originalMaxLength;
+        private readonly Stream _originalStream;
+        private readonly bool _error;
+        private readonly Encoding? _textEncoding;
+        private int _blockNumber;
+        private readonly PipelineMessageLogger _messageLogger;
+
+        public LoggingStream(PipelineMessageLogger messageLogger, string requestId, int maxLoggedBytes, Stream originalStream, bool error, Encoding? textEncoding)
+        {
+            // Should only wrap non-seekable streams
+            Debug.Assert(!originalStream.CanSeek);
+            _requestId = requestId;
+            _maxLoggedBytes = maxLoggedBytes;
+            _originalMaxLength = maxLoggedBytes;
+            _originalStream = originalStream;
+            _error = error;
+            _textEncoding = textEncoding;
+            _messageLogger = messageLogger;
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            return _originalStream.Seek(offset, origin);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            var result = _originalStream.Read(buffer, offset, count);
+
+            var countToLog = result;
+            DecrementLength(ref countToLog);
+            LogBuffer(buffer, offset, countToLog, false);
+
+            return result;
+        }
+
+        private void LogBuffer(byte[] buffer, int offset, int length, bool async)
+        {
+            if (length == 0 || buffer == null)
+            {
+                return;
+            }
+
+            var logLength = Math.Min(length, _originalMaxLength);
+
+            byte[] bytes;
+            if (length == logLength && offset == 0)
+            {
+                bytes = buffer;
+            }
+            else
+            {
+                bytes = new byte[logLength];
+                Buffer.BlockCopy(buffer, offset, bytes, 0, logLength);
+            }
+
+            if (_error)
+            {
+                _messageLogger.LogErrorResponseContentBlock(_requestId, _blockNumber, bytes, _textEncoding);
+            }
+            else
+            {
+                _messageLogger.LogResponseContentBlock(_requestId, _blockNumber, bytes, _textEncoding);
+            }
+
+            _blockNumber++;
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+#pragma warning disable CA1835 // ReadAsync(Memory<>) overload is not available in netstandard2.0
```

Comment:
Do we need a platform-if here?   What do we do if we're running in a `netstandard2.0` scenario?

Reply:
This one sounds more concerning than it is:
[CA1835: Prefer the memory-based overloads of ReadAsync/WriteAsync methods in stream-based classes](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1835) it's just less efficient. I'll work on getting the more efficient code path for the newer targets